service/src/main/java/uk/gov/hmcts/reform/fpl/service/MessageJudgeService.java

```diff
@@ -81,23 +125,69 @@
             C2DocumentBundle selectedC2DocumentBundle = caseData.getC2DocumentBundleByUUID(selectedC2Id);
 
             judicialMessageBuilder.relatedDocuments(selectedC2DocumentBundle.getAllC2DocumentReferences());
+            judicialMessageBuilder.relatedDocumentFileNames(selectedC2DocumentBundle.getAllC2DocumentFileNames());
+            judicialMessageBuilder.isRelatedToC2(YES);
         }
 
         judicialMessages.add(element(identityService.generateId(), judicialMessageBuilder.build()));
         return judicialMessages;
     }
 
+    public List<Element<JudicialMessage>> replyToJudicialMessage(CaseData caseData) {
+        List<Element<JudicialMessage>> judicialMessages = caseData.getJudicialMessages();
+        MessageJudgeEventData messageJudgeEventData = caseData.getMessageJudgeEventData();
+        JudicialMessage judicialMessageReply = messageJudgeEventData.getJudicialMessageReply();
+
+        UUID selectedJudicialMessageId = getDynamicListSelectedValue(
+            caseData.getMessageJudgeEventData().getJudicialMessageDynamicList(), mapper
+        );
+
+        return judicialMessages.stream()
+            .map(judicialMessageElement -> {
+                if (selectedJudicialMessageId.equals(judicialMessageElement.getId())) {
+
+                    JudicialMessage judicialMessage = judicialMessageElement.getValue();
+
+                    String sender = userService.getUserEmail();
+
+                    JudicialMessage updatedMessage = judicialMessage.toBuilder()
+                        .updatedTime(time.now())
+                        .sender(sender) // Get the email of the current user
+                        .recipient(judicialMessage.getSender()) // Get the sender of the previous message
+                        .messageHistory(String.join("\n \n", List.of(
```

Comment:
I would extract this to a method, something like
```java
    private String buildMessageHistory(String message, String sender) {
        return buildMessageHistory(message, "", sender);
    }

    private String buildMessageHistory(JudicialMessage reply, JudicialMessage previousMessage, String sender) {
        return buildMessageHistory(reply.getLatestMessage(), previousMessage.getMessageHistory(), sender);
    }

    private String buildMessageHistory(String message, String history, String sender) {
        String formattedLatestMessage = String.format("%s - %s", sender, message);

        if (history.isBlank()) {
            return formattedLatestMessage;
        }

        return String.join("\n \n", history, formattedLatestMessage);
    }
```
with those overloads you can then call the method in the `addNewJudicialMessage` method as well

**Reply**
```java
                    JudicialMessage updatedMessage = judicialMessage.toBuilder()
                        .updatedTime(time.now())
                        .sender(sender) // Get the email of the current user
                        .recipient(judicialMessage.getSender()) // Get the sender of the previous message
                        .messageHistory(buildMessageHistory(judicialMessageReply, judicialMessage, sender))
                        .latestMessage(judicialMessageReply.getLatestMessage())
                        .build();
```
**New**
```java
        JudicialMessage.JudicialMessageBuilder<?, ?> judicialMessageBuilder = JudicialMessage.builder()
            .sender(sender)
            .recipient(judicialMessageMetaData.getRecipient())
            .requestedBy(judicialMessageMetaData.getRequestedBy())
            .latestMessage(latestMessage)
            .messageHistory(buildMessageHistory(latestMessage, sender))
            .updatedTime(time.now())
            .dateSent(formatLocalDateTimeBaseUsingFormat(time.now(), DATE_TIME_AT))
            .urgency(judicialMessageMetaData.getUrgency())
            .status(OPEN);
```